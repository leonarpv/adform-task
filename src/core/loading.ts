export enum LoadingStatus {
    Idle = 'Idle',
    Pending = 'Pending',
    Succeeded = 'Succeeded',
    Failed = 'Failed',
}
