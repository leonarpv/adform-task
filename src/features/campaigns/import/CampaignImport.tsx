import React, { ChangeEvent, FormEvent, useState } from 'react'
import Modal from 'react-modal'
import Button from 'components/button/Button'
import TextArea from 'components/fields/text-area/TextArea'
import './campaignImport.scss'

type Props = {
    onImport: (input: string) => void
}

const CampaignImport = ({ onImport }: Props) => {
    const [isOpen, setIsOpen] = useState<boolean>(false)
    const [value, setValue] = useState<string>('')

    const openModal = () => {
        setIsOpen(true)
    }

    const closeModal = () => {
        setIsOpen(false)
    }

    const handleChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
        setValue(e.target.value)
    }

    const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault()

        onImport(value)
    }

    return (
        <>
            <Button onClick={openModal}>Import Campaigns</Button>
            <Modal
                appElement={document.body}
                isOpen={isOpen}
                onRequestClose={closeModal}
                contentLabel="Example Modal"
            >
                <form className="campaign-import" onSubmit={handleSubmit}>
                    <TextArea
                        label="Enter Campaign Data"
                        name="message"
                        value={value}
                        onChange={handleChange}
                    />
                    <div className="campaign-import__actions">
                        <Button type="submit">Import</Button>
                        <Button type="button" onClick={closeModal} secondary>
                            Cancel
                        </Button>
                    </div>
                </form>
            </Modal>
        </>
    )
}

export default CampaignImport
