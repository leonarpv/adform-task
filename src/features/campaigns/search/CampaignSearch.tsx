import React, { ChangeEvent, ReactNode } from 'react'
import { CampaignFilters } from '../types'
import Datepicker from 'components/fields/datepicker/Datepicker'
import Input from 'components/fields/input/Input'
import dayjs from 'dayjs'
import './campaignSearch.scss'

type Props = {
    children?: ReactNode
    filters?: CampaignFilters
    onChange: (search: CampaignFilters) => void
}

const CampaignSearch = ({ children, filters, onChange }: Props) => {
    const handleStartDateChange = (date: Date | null) => {
        onChange({ ...filters, startDate: date?.toString() })
    }

    const handleEndDateChange = (date: Date | null) => {
        onChange({ ...filters, endDate: date?.toString() })
    }

    const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
        onChange({ ...filters, name: e.target.value })
    }

    const startDate = filters?.startDate
        ? dayjs(filters.startDate).toDate()
        : undefined

    const endDate = filters?.endDate
        ? dayjs(filters.endDate).toDate()
        : undefined

    const maxStartDate = filters?.endDate
        ? dayjs(filters.endDate).subtract(1, 'day').toDate()
        : undefined

    const minEndDate = filters?.startDate
        ? dayjs(filters.startDate).add(1, 'day').toDate()
        : undefined

    return (
        <div className="campaign-search">
            <div className="campaign-search__inputs">
                <div className="campaign-search__date-pickers">
                    {children}
                    <Datepicker
                        label="Start Date"
                        isClearable
                        maxDate={maxStartDate}
                        selected={startDate}
                        onChange={handleStartDateChange}
                    />
                    <Datepicker
                        label="End Date"
                        isClearable
                        minDate={minEndDate}
                        selected={endDate}
                        onChange={handleEndDateChange}
                    />
                </div>

                <Input
                    label="Search By Name"
                    value={filters?.name || ''}
                    type="text"
                    onChange={handleNameChange}
                />
            </div>
        </div>
    )
}

export default CampaignSearch
