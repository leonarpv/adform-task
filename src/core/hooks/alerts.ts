import { useEffect } from 'react'
import { toast } from 'react-toastify'
import { LoadingStatus } from '../loading'

type Config = {
    status: LoadingStatus
    message?: string
}

export const useFailedLoadingToast = ({ message, status }: Config) =>
    useEffect(() => {
        if (status === LoadingStatus.Failed) {
            toast(message, { type: 'error' })
        }
    }, [message, status])

export const useSuccessLoadingToast = ({ message, status }: Config) =>
    useEffect(() => {
        if (status === LoadingStatus.Succeeded) {
            toast(message, { type: 'success' })
        }
    }, [message, status])
