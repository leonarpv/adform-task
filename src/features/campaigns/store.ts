import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { getCampaignUsers } from './api'
import {
    Campaign,
    CampaignFilters,
    CampaignUser,
    CampaignWithUser,
} from './types'
import { LoadingStatus } from 'core/loading'
import { filterCampaigns, validateCampaignData } from './utils'

type CampaignUserMap = { [key: number]: CampaignUser }

interface CampaignsState {
    campaigns: CampaignWithUser[]
    campaignsFiltered: CampaignWithUser[]
    campaignsStatus: LoadingStatus
    campaignUsers: CampaignUserMap
    campaignUsersLoaded: boolean
    campaignUsersStatus: LoadingStatus
    campaignValidationError?: string
    filters?: CampaignFilters
}

export const fetchCampaignUsers = createAsyncThunk<
    CampaignUserMap,
    void,
    { state: CampaignsState }
>('campaigns/fetchCampaignUsers', async (_, { getState }) => {
    const { campaignUsers, campaignUsersLoaded } = getState()

    if (campaignUsersLoaded) {
        return campaignUsers
    }

    const { data } = await getCampaignUsers()

    const userMap = data.reduce(
        (acc: CampaignUserMap, u: CampaignUser) =>
            Object.assign(acc, { [u.id]: u }),
        {}
    )

    return userMap
})

export const importCampaigns = createAsyncThunk<
    CampaignWithUser[],
    unknown,
    { state: CampaignsState }
>(
    'campaigns/importCampaigns',
    async (input: unknown, { dispatch, getState }) => {
        const campaigns = typeof input === 'string' ? JSON.parse(input) : input

        const validationError = validateCampaignData(campaigns)

        if (validationError) {
            throw new Error(validationError)
        }

        await dispatch(fetchCampaignUsers())

        const { campaignUsers } = getState()

        const campaignsWithUsers = (campaigns as Campaign[]).map((c) => ({
            ...c,
            user: campaignUsers[c.userId] || null,
        }))

        return campaignsWithUsers
    }
)

const initialState = {
    campaigns: [],
    campaignsFiltered: [],
    campaignsStatus: LoadingStatus.Idle,
    campaignUsers: {},
    campaignUsersLoaded: false,
    campaignUsersStatus: LoadingStatus.Idle,
    campaignValidationError: undefined,
    filters: undefined,
} as CampaignsState

export const campaignsSlice = createSlice({
    name: 'campaigns',
    initialState,
    reducers: {
        setCampaignFilters: (state, action) => {
            state.filters = action.payload
            state.campaignsFiltered = filterCampaigns(
                state.campaigns,
                state.filters
            )
        },
    },
    extraReducers: {
        [fetchCampaignUsers.pending.type]: (state) => {
            state.campaignUsersStatus = LoadingStatus.Pending
        },
        [fetchCampaignUsers.fulfilled.type]: (state, action) => {
            state.campaignUsers = action.payload
            state.campaignUsersLoaded = true
            state.campaignUsersStatus = LoadingStatus.Succeeded
        },
        [fetchCampaignUsers.rejected.type]: (state) => {
            state.campaignUsersStatus = LoadingStatus.Failed
        },
        [importCampaigns.pending.type]: (state) => {
            state.campaignsStatus = LoadingStatus.Pending
            state.campaignValidationError = undefined
        },
        [importCampaigns.fulfilled.type]: (state, action) => {
            const campaigns = [...state.campaigns, ...action.payload]

            state.campaigns = campaigns
            state.campaignsStatus = LoadingStatus.Succeeded
            state.campaignsFiltered = filterCampaigns(campaigns, state.filters)
        },
        [importCampaigns.rejected.type]: (state, action) => {
            state.campaignsStatus = LoadingStatus.Failed
            state.campaignValidationError = action.error?.message
        },
    },
})

export const { setCampaignFilters } = campaignsSlice.actions

export const selectCampaigns = (state: CampaignsState) => state
