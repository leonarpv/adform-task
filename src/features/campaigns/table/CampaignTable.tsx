import React from 'react'
import DataTable from 'components/data-table/DataTable'
import { ColumnDefinition } from 'components/data-table/types'
import { CampaignWithUser } from '../types'
import { currencyFormatter } from 'core/currencyFormatter'
import { dateFormatter } from 'core/dateFormatter'
import { LoadingStatus } from 'core/loading'
import CampaignStatus from '../status/CampaignStatus'
import dayjs from 'dayjs'

type Props = {
    campaigns: CampaignWithUser[]
    campaignsStatus: LoadingStatus
}

export const columns: ColumnDefinition<CampaignWithUser>[] = [
    {
        title: 'Name',
        accessor: (c) => c.name,
    },
    {
        title: 'User Name',
        accessor: (c) => (c.user ? c.user.name : 'Unknown user'),
    },
    {
        title: 'Start Date',
        accessor: (c) => dateFormatter(c.startDate),
    },
    {
        title: 'End Date',
        accessor: (c) => dateFormatter(c.endDate),
    },
    {
        title: 'Active',
        accessor: (c) => {
            const startDate = dayjs(c.startDate)
            const endDate = dayjs(c.endDate)
            const currentDate = dayjs()

            const active =
                currentDate.isAfter(startDate) && currentDate.isBefore(endDate)

            return <CampaignStatus active={active} />
        },
    },
    {
        title: 'Budget',
        accessor: (c) => currencyFormatter(c.Budget),
    },
]

const CampaignTable = ({ campaigns, campaignsStatus }: Props) => {
    const emptyText =
        campaignsStatus === LoadingStatus.Pending
            ? 'Loading..'
            : 'No data to display'

    return (
        <DataTable columns={columns} emptyText={emptyText} source={campaigns} />
    )
}

export default CampaignTable
