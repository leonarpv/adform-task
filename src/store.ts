import { useDispatch } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import { campaignsSlice } from './features/campaigns/store'

export const store = configureStore({
    reducer: campaignsSlice.reducer,
})

export type AppDispatch = typeof store.dispatch
export const useAppDispatch = () => useDispatch<AppDispatch>()
