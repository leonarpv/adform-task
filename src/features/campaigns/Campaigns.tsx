import React from 'react'
import { useFailedLoadingToast, useSuccessLoadingToast } from 'core/loading'
import CampaignTable from './table/CampaignTable'
import CampaignSearch from './search/CampaignSearch'
import { Campaign, CampaignFilters } from './types'
import { importCampaigns, selectCampaigns, setCampaignFilters } from './store'
import { useAppDispatch } from 'store'
import { useSelector } from 'react-redux'
import CampaignImport from './import/CampaignImport'

declare global {
    interface Window {
        AddCampaigns: (campaigns: Campaign[]) => void
    }
}

const Campaigns = () => {
    const dispatch = useAppDispatch()

    const {
        campaignsStatus,
        campaignUsersStatus,
        campaignsFiltered,
        campaignValidationError,
        filters,
    } = useSelector(selectCampaigns)

    useSuccessLoadingToast({
        status: campaignsStatus,
        message: 'Campaign Data Imported',
    })
    useFailedLoadingToast({
        status: campaignsStatus,
        message: campaignValidationError,
    })
    useFailedLoadingToast({
        status: campaignUsersStatus,
        message: 'User Fetch Failed',
    })

    window.AddCampaigns = (campaigns: Campaign[]) => {
        dispatch(importCampaigns(campaigns))
    }

    const handleChange = (filters: CampaignFilters) => {
        dispatch(setCampaignFilters(filters))
    }

    const handleImport = (campaignString: string) => {
        dispatch(importCampaigns(campaignString))
    }

    return (
        <>
            <CampaignSearch filters={filters} onChange={handleChange}>
                <CampaignImport onImport={handleImport} />
            </CampaignSearch>
            <CampaignTable
                campaigns={campaignsFiltered}
                campaignsStatus={campaignsStatus}
            />
        </>
    )
}

export default Campaigns
