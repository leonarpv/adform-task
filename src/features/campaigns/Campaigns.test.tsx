import { shallow } from 'enzyme'
import Campaigns from './Campaigns'
import React from 'react'
import * as alerts from 'core/hooks/alerts'
import * as store from 'store'
import * as reactRedux from 'react-redux'
import CampaignSearch from './search/CampaignSearch'
import CampaignImport from './import/CampaignImport'
import CampaignTable from './table/CampaignTable'
import { CampaignsState, selectCampaigns } from './store'
import { Mock } from 'ts-mockery'
import { LoadingStatus } from '../../core/loading'

jest.mock('core/hooks/alerts', () => ({
    useFailedLoadingToast: jest.fn(),
    useSuccessLoadingToast: jest.fn(),
}))

jest.mock('store', () => ({
    useAppDispatch: jest.fn(),
}))

jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
}))

describe('Campaigns', () => {
    beforeEach(() => jest.clearAllMocks())

    it('should render properly', () => {
        const state = Mock.of<CampaignsState>({})

        jest.spyOn(alerts, 'useFailedLoadingToast')
        jest.spyOn(alerts, 'useSuccessLoadingToast')
        jest.spyOn(store, 'useAppDispatch')
        jest.spyOn(reactRedux, 'useSelector').mockReturnValue(state)

        const wrapper = shallow(<Campaigns />)

        expect(wrapper).toMatchSnapshot()
    })

    it('should initialize component', () => {
        const state = Mock.of<CampaignsState>({
            campaignsStatus: LoadingStatus.Failed,
            campaignUsersStatus: LoadingStatus.Succeeded,
            campaignValidationError: 'campaignValidationError',
        })

        const useFailedLoadingToast = jest.spyOn(
            alerts,
            'useFailedLoadingToast'
        )
        const useSuccessLoadingToast = jest.spyOn(
            alerts,
            'useSuccessLoadingToast'
        )
        const useAppDispatch = jest.spyOn(store, 'useAppDispatch')
        const useSelector = jest
            .spyOn(reactRedux, 'useSelector')
            .mockReturnValue(state)

        const wrapper = shallow(<Campaigns />)

        expect(useFailedLoadingToast).toHaveBeenCalledTimes(2)
        expect(useFailedLoadingToast).toHaveBeenNthCalledWith(1, {
            status: state.campaignsStatus,
            message: state.campaignValidationError,
        })
        expect(useFailedLoadingToast).toHaveBeenNthCalledWith(2, {
            status: state.campaignUsersStatus,
            message: 'User Fetch Failed',
        })

        expect(useSuccessLoadingToast).toHaveBeenCalledTimes(1)
        expect(useSuccessLoadingToast).toHaveBeenCalledWith({
            status: state.campaignsStatus,
            message: 'Campaign Data Imported',
        })

        expect(useAppDispatch).toHaveBeenCalledTimes(1)

        expect(useSelector).toHaveBeenCalledTimes(1)
        expect(useSelector).toHaveBeenCalledWith(selectCampaigns)

        expect(wrapper.find(CampaignSearch).exists()).toBe(true)
        expect(wrapper.find(CampaignImport).exists()).toBe(true)
        expect(wrapper.find(CampaignTable).exists()).toBe(true)
    })
})
