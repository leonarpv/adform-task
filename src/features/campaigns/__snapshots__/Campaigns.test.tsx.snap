// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Campaigns should render properly 1`] = `ShallowWrapper {}`;
